File: src/smartest/internal/smartest_discovery_ffi.erl
%% SPDX-FileCopyrightText: 2026 gleam_mutants contributors
%% SPDX-License-Identifier: MIT OR Apache-2.0

-module(smartest_discovery_ffi).
-export([main/0]).

main() ->
    reset_protocol_state(),
    case load_selection() of
        {error, Reason} ->
            io:format("FAIL test selection protocol~n~ts~n", [Reason]),
            erlang:halt(2);
        Selection -> erlang:put(smartest_selection, Selection)
    end,
    Package = package_name(),
    Filter = env(<<"SMARTEST_FILTER">>),
    Files = filelib:wildcard("test/**/*.{erl,gleam}"),
    Modules = lists:sort([module_from_path(File) || File <- Files]),
    {Passed0, Failed0, Bindings, ManifestComplete, Results} = lists:foldl(
      fun(Module, Counts) -> run_module(Package, Module, Filter, Counts) end,
      {0, 0, [], true, []},
      Modules
    ),
    {Passed1, Failed1} = validate_selected(Passed0, Failed0),
    {Passed2, Failed2} = write_evidence_report(Results, Passed1, Failed1),
    {Passed3, Failed3} = maybe_write_manifest(
      Filter, Bindings, ManifestComplete, Passed2, Failed2),
    {Passed, Failed} = maybe_write_impact(
      Filter, ManifestComplete, Passed3, Failed3),
    io:format("~n~B passed, ~B failed~n", [Passed, Failed]),
    case Failed of
        0 -> erlang:halt(0);
        _ -> erlang:halt(1)
    end.

run_module(Package, Module, Filter, Counts) ->
    case code:ensure_loaded(Module) of
        {module, Module} ->
            Exports = lists:sort([
              Function || {Function, 0} <- Module:module_info(exports),
                          is_test_name(Function),
                          selected(Module, Function, Filter)
            ]),
            lists:foldl(
              fun(Function, Acc) -> run_export(Package, Module, Function, Acc) end,
              Counts,
              Exports
            );
        {error, Reason} ->
            io:format("FAIL ~tp could not load: ~tp~n", [Module, Reason]),
            mark_incomplete(increment_failed(Counts))
    end.

run_export(Package, Module, Function, Counts) ->
    Root = test_id(Package, Module, Function),
    case selected_export(Root) of
        false -> Counts;
        true -> run_selected_export(Package, Module, Function, Root, Counts)
    end.

run_selected_export(Package, Module, Function, Root, Counts) ->
    try smartest_runtime_ffi:with_test_context(
          Root, fun() -> apply(Module, Function, []) end) of
        Value when is_tuple(Value), tuple_size(Value) =:= 3,
                   element(1, Value) =:= test ->
            Entry = 'smartest@runner':entry(
              Package,
              atom_to_binary(Module, utf8),
              atom_to_binary(Function, utf8),
              Value
            ),
            Ids = case protocol_enabled() of
                true -> ['smartest@evidence':test_id_to_string(Id)
                         || Id <- 'smartest@runner':test_ids(Entry)];
                false -> []
            end,
            add_descriptors(Ids, <<"smartest-leaf">>),
            Bindings = 'smartest@runner':generator_bindings(Entry),
            Reports = case selection() of
                all -> 'smartest@runner':run_entry(Entry, workspace_options());
                {selected, Selectors} ->
                    Matching = [Id || Id <- Ids, lists:member(Id, Selectors)],
                    mark_matched(Matching),
                    'smartest@runner':run_entry_selected(
                      Entry, workspace_options(), Matching)
            end,
            lists:foldl(fun report/2, add_bindings(Counts, Bindings), Reports);
        _LegacyValue ->
            add_descriptor(Root, <<"legacy-export">>),
            mark_matched([Root]),
            io:put_chars("."),
            Legacy = 'smartest@runner':legacy_result(
              Package,
              atom_to_binary(Module, utf8),
              atom_to_binary(Function, utf8),
              true,
              <<"opaque legacy test">>
            ),
            increment_passed(add_result(Counts, Legacy))
    catch
        Class:Reason:Stack ->
            Message = smartest_runtime_ffi:format_failure(Class, Reason, Stack),
            io:format(
              "~nFAIL ~ts/~ts~n~ts~n",
              [atom_to_binary(Module, utf8), atom_to_binary(Function, utf8),
               Message]
            ),
            add_descriptor(Root, <<"legacy-export">>),
            mark_matched([Root]),
            Legacy = 'smartest@runner':legacy_result(
              Package,
              atom_to_binary(Module, utf8),
              atom_to_binary(Function, utf8),
              false,
              Message
            ),
            mark_incomplete(increment_failed(add_result(Counts, Legacy)))
    end.

report(Result, Counts) ->
    WithResult = add_result(Counts, Result),
    case 'smartest@runner':succeeded(Result) of
        true ->
            io:put_chars("."),
            increment_passed(WithResult);
        false ->
            io:format("~n~ts", ['smartest@runner':render_result(Result)]),
            increment_failed(WithResult)
    end.

increment_passed({Passed, Failed, Bindings, Complete, Results}) ->
    {Passed + 1, Failed, Bindings, Complete, Results}.
increment_failed({Passed, Failed, Bindings, Complete, Results}) ->
    {Passed, Failed + 1, Bindings, Complete, Results}.

add_bindings({Passed, Failed, Existing, Complete, Results}, Bindings) ->
    {Passed, Failed, Existing ++ Bindings, Complete, Results}.

add_result({Passed, Failed, Bindings, Complete, Results}, Result) ->
    {Passed, Failed, Bindings, Complete, Results ++ [Result]}.

mark_incomplete({Passed, Failed, Bindings, _, Results}) ->
    {Passed, Failed, Bindings, false, Results}.

write_evidence_report(Results, Passed, Failed) ->
    {ok, Directory} = file:get_cwd(),
    Root = unicode:characters_to_binary(Directory),
    case 'smartest@report':write(Root, {report, Results}) of
        {ok, _} -> {Passed, Failed};
        {error, Reason} ->
            io:format("~nFAIL evidence report~n~ts~n", [Reason]),
            {Passed, Failed + 1}
    end.

maybe_write_manifest(Filter, Bindings, Complete, Passed, Failed) ->
    case selection() of
        {selected, _} -> {Passed, Failed};
        all -> maybe_write_full_manifest(Filter, Bindings, Complete, Passed, Failed)
    end.

maybe_write_full_manifest(Filter, _Bindings, _Complete, Passed, Failed)
  when Filter =/= <<>> -> {Passed, Failed};
maybe_write_full_manifest(_Filter, _Bindings, false, Passed, Failed) ->
    io:format("~nGenerator manifest was not replaced because discovery was incomplete.~n"),
    {Passed, Failed};
maybe_write_full_manifest(_Filter, Bindings, true, Passed, Failed) ->
    {ok, Directory} = file:get_cwd(),
    Root = unicode:characters_to_binary(Directory),
    case 'smartest@storage':write_generator_manifest(Root, Bindings) of
        {ok, _} -> {Passed, Failed};
        {error, Reason} ->
            io:format("~nFAIL generator manifest~n~ts~n", [Reason]),
            {Passed, Failed + 1}
    end.

is_test_name(Function) ->
    lists:suffix("_test", atom_to_list(Function)).

selected(_Module, _Function, <<>>) -> true;
selected(Module, Function, Filter) ->
    Name = <<(atom_to_binary(Module, utf8))/binary, "/",
             (atom_to_binary(Function, utf8))/binary>>,
    binary:match(Name, Filter) =/= nomatch.

module_from_path(Path) ->
    Relative0 = lists:flatten(string:replace(Path, "\\", "/", all)),
    Relative1 = case lists:prefix("test/", Relative0) of
        true -> lists:nthtail(length("test/"), Relative0);
        false -> Relative0
    end,
    Relative2 = filename:rootname(Relative1),
    list_to_atom(lists:flatten(string:replace(Relative2, "/", "@", all))).

package_name() ->
    case file:read_file("gleam.toml") of
        {ok, Source} ->
            case re:run(
              Source,
              <<"(?m)^\\s*name\\s*=\\s*\"([a-z][a-z0-9_]*)\"">>,
              [{capture, [1], binary}]
            ) of
                {match, [Name]} -> Name;
                _ -> <<"unknown-package">>
            end;
        _ -> <<"unknown-package">>
    end.

env(Name) ->
    case os:getenv(unicode:characters_to_list(Name)) of
        false -> <<>>;
        Value -> unicode:characters_to_binary(Value)
    end.

workspace_options() ->
    {ok, Directory} = file:get_cwd(),
    Root = unicode:characters_to_binary(Directory),
    Now = erlang:system_time(millisecond),
    case env(<<"SMARTEST_REPLAY_ID">>) of
        <<>> -> 'smartest@runner':workspace_options(Root, Now);
        Id -> 'smartest@runner':replay_options(Root, Id, Now)
    end.

%% Adaptive protocol ---------------------------------------------------------

reset_protocol_state() ->
    erlang:put(smartest_descriptors, []),
    erlang:put(smartest_matched_selectors, []),
    case ets:whereis(gleam_mutants_test_impact) of
        undefined -> ok;
        Table -> ets:delete(Table)
    end,
    %% The discovery process owns the table for the whole suite. Test bodies
    %% may run in short-lived child processes, so lazily creating it there
    %% would discard their observations when the child exits.
    case env(<<"GLEAM_MUTANTS_TEST_IMPACT_FILE">>) of
        <<>> -> ok;
        _ -> ets:new(gleam_mutants_test_impact, [named_table, public, set])
    end,
    ok.

protocol_enabled() ->
    env(<<"GLEAM_MUTANTS_TEST_IMPACT_FILE">>) =/= <<>> orelse
      env(<<"GLEAM_MUTANTS_TEST_SELECTION_FILE">>) =/= <<>>.

selection() ->
    case erlang:get(smartest_selection) of
        undefined -> all;
        Value -> Value
    end.

load_selection() ->
    case env(<<"GLEAM_MUTANTS_TEST_SELECTION_FILE">>) of
        <<>> -> all;
        Path ->
            case valid_protocol_path(Path) of
                false -> {error, <<"selection file must be below .gleam_mutants">>};
                true -> decode_selection(Path)
            end
    end.

decode_selection(Path) ->
    try
        {ok, Source} = file:read_file(Path),
        Document = json:decode(Source),
        #{<<"schema_version">> := 1,
          <<"runner">> := <<"smartest">>,
          <<"runtime">> := Runtime,
          <<"selectors">> := Selectors} = Document,
        true = Runtime =:= runtime_name(),
        true = is_list(Selectors),
        true = lists:all(fun is_binary/1, Selectors),
        true = length(Selectors) =:= length(lists:usort(Selectors)),
        {selected, Selectors}
    catch
        _:_ -> {error, <<"invalid or incompatible test selection file">>}
    end.

valid_protocol_path(Path0) ->
    {ok, Directory} = file:get_cwd(),
    Path = filename:absname(unicode:characters_to_list(Path0)),
    Base = filename:join(filename:absname(Directory), ".gleam_mutants"),
    Path =:= Base orelse lists:prefix(Base ++ "/", Path)
      orelse lists:prefix(Base ++ "\\", Path).

runtime_name() ->
    case env(<<"GLEAM_MUTANTS_RUNTIME">>) of
        <<>> -> <<"erlang">>;
        Runtime -> Runtime
    end.

test_id(Package, Module, Function) ->
    Id = 'smartest@evidence':test_id(
      Package, atom_to_binary(Module, utf8), atom_to_binary(Function, utf8)),
    'smartest@evidence':test_id_to_string(Id).

selected_export(Root) ->
    case selection() of
        all -> true;
        {selected, Selectors} ->
            Prefix = binary_to_list(<<Root/binary, "/">>),
            lists:any(fun(Selector) ->
              Selector =:= Root orelse lists:prefix(Prefix, binary_to_list(Selector))
            end, Selectors)
    end.

add_descriptors(Ids, Kind) ->
    lists:foreach(fun(Id) -> add_descriptor(Id, Kind) end, Ids).

add_descriptor(Id, Kind) ->
    case protocol_enabled() of
        false -> ok;
        true ->
            Existing = erlang:get(smartest_descriptors),
            erlang:put(smartest_descriptors, [{Id, Id, Kind} | Existing]),
            ok
    end.

mark_matched(Ids) ->
    case protocol_enabled() of
        false -> ok;
        true ->
            Existing = erlang:get(smartest_matched_selectors),
            erlang:put(smartest_matched_selectors, Ids ++ Existing),
            ok
    end.

validate_selected(Passed, Failed) ->
    case selection() of
        all -> {Passed, Failed};
        {selected, Selectors} ->
            Matched = lists:usort(erlang:get(smartest_matched_selectors)),
            Unknown = [Selector || Selector <- Selectors,
                                    not lists:member(Selector, Matched)],
            case Unknown of
                [] -> {Passed, Failed};
                _ ->
                    io:format("~nFAIL unknown test selector~n~tp~n", [Unknown]),
                    {Passed, Failed + 1}
            end
    end.

maybe_write_impact(_Filter, Complete0, Passed, Failed) ->
    case env(<<"GLEAM_MUTANTS_TEST_IMPACT_FILE">>) of
        <<>> -> {Passed, Failed};
        Path ->
            %% A runner filter is part of the configured test command, so its
            %% selected tests are the complete suite that a later full-suite
            %% confirmation will execute as well.
            Complete = Complete0 andalso selection() =:= all,
            case write_impact(Path, Complete) of
                ok -> {Passed, Failed};
                {error, Reason} ->
                    io:format("~nFAIL test impact manifest~n~ts~n", [Reason]),
                    {Passed, Failed + 1}
            end
    end.

write_impact(Path, Complete) ->
    case valid_protocol_path(Path) of
        false -> {error, <<"impact file must be below .gleam_mutants">>};
        true ->
            Descriptors = lists:reverse(erlang:get(smartest_descriptors)),
            Hits = case ets:whereis(gleam_mutants_test_impact) of
                undefined -> [];
                Table -> ets:tab2list(Table)
            end,
            Tests = [#{<<"selector">> => Selector,
                       <<"test_id">> => TestId,
                       <<"kind">> => Kind}
                     || {Selector, TestId, Kind} <- Descriptors],
            Reaches = [#{<<"test_id">> => TestId,
                         <<"mutant_ids">> => lists:sort([
                           Mutant || {{HitTest, Mutant}} <- Hits,
                                     HitTest =:= TestId])}
                       || {_Selector, TestId, _Kind} <- Descriptors],
            Document = #{<<"schema_version">> => 1,
                         <<"runner">> => <<"smartest">>,
                         <<"runtime">> => runtime_name(),
                         <<"complete">> => Complete,
                         <<"tests">> => Tests,
                         <<"reaches">> => Reaches},
            atomic_write(Path, iolist_to_binary(json:encode(Document)))
    end.

atomic_write(Path0, Contents) ->
    Path = unicode:characters_to_list(Path0),
    Temporary = Path ++ ".tmp-" ++ integer_to_list(erlang:unique_integer([positive])),
    ok = filelib:ensure_dir(Path),
    case file:write_file(Temporary, Contents, [binary, sync]) of
        ok ->
            case file:rename(Temporary, Path) of
                ok -> ok;
                {error, Reason} ->
                    _ = file:delete(Temporary),
                    {error, unicode:characters_to_binary(io_lib:format("~tp", [Reason]))}
            end;
        {error, Reason} ->
            _ = file:delete(Temporary),
            {error, unicode:characters_to_binary(io_lib:format("~tp", [Reason]))}
    end.

File: src/smartest/internal/smartest_discovery_ffi.mjs
// SPDX-FileCopyrightText: 2026 gleam_mutants contributors
// SPDX-License-Identifier: MIT OR Apache-2.0

import { Metadata, metadata } from "./plan.mjs";
import { with_test_context } from "./smartest_runtime_ffi.mjs";
import { toList } from "../../gleam.mjs";
import * as evidence from "../evidence.mjs";
import * as runner from "../runner.mjs";
import * as storage from "../storage.mjs";
import * as evidenceReport from "../report.mjs";

async function readDirectory(directory) {
  if (globalThis.Deno) {
    const names = [];
    for await (const entry of Deno.readDir(directory)) names.push(entry.name);
    return names.sort();
  }
  const { readdir } = await import("node:fs/promises");
  return (await readdir(directory)).sort();
}

async function* gleamFiles(directory) {
  for (const entry of await readDirectory(directory)) {
    const candidate = `${directory}/${entry}`;
    if (candidate.endsWith(".gleam")) {
      yield candidate;
      continue;
    }
    try {
      yield* gleamFiles(candidate);
    } catch (_) {
      // A non-directory entry is not a Gleam test module.
    }
  }
}

async function readText(path) {
  if (globalThis.Deno) return Deno.readTextFile(path);
  const { readFile } = await import("node:fs/promises");
  return String(await readFile(path));
}

async function packageName() {
  const source = await readText("gleam.toml");
  return source.match(/^\s*name\s*=\s*"([a-z][a-z0-9_]*)"/m)?.[1]
    ?? "unknown-package";
}

function selected(moduleName, functionName, filter) {
  return !filter || `${moduleName}/${functionName}`.includes(filter);
}

function environment(name) {
  try {
    if (globalThis.Deno) return Deno.env.get(name) ?? "";
  } catch (_) {
    return "";
  }
  return globalThis.process?.env?.[name] ?? "";
}

function print(value) {
  if (globalThis.Deno) Deno.stdout.writeSync(new TextEncoder().encode(value));
  else globalThis.process.stdout.write(value);
}

function finish(code) {
  if (globalThis.Deno) Deno.exit(code);
  globalThis.process.exit(code);
}

const impactSymbol = Symbol.for("gleam-mutants.test-impact");

function runtimeName() {
  return environment("GLEAM_MUTANTS_RUNTIME")
    || (globalThis.Deno ? "deno" : globalThis.Bun ? "bun" : "node");
}

async function validProtocolPath(candidate) {
  const path = await import("node:path");
  const absolute = path.resolve(candidate);
  const base = path.resolve(".gleam_mutants");
  return absolute.startsWith(`${base}${path.sep}`);
}

async function loadSelection() {
  const file = environment("GLEAM_MUTANTS_TEST_SELECTION_FILE");
  if (!file) return null;
  if (!(await validProtocolPath(file))) {
    throw new Error("selection file must be below .gleam_mutants");
  }
  const value = JSON.parse(await readText(file));
  if (value?.schema_version !== 1
    || value?.runner !== "smartest"
    || value?.runtime !== runtimeName()
    || !Array.isArray(value?.selectors)
    || !value.selectors.every(selector => typeof selector === "string")
    || new Set(value.selectors).size !== value.selectors.length) {
    throw new Error("invalid or incompatible test selection file");
  }
  return value.selectors;
}

function rootTestId(package_, moduleName, functionName) {
  return evidence.test_id_to_string(
    evidence.test_id(package_, moduleName, functionName),
  );
}

function selectedExport(root, selectors) {
  return selectors === null
    || selectors.some(selector => selector === root || selector.startsWith(`${root}/`));
}

async function atomicWrite(target, source) {
  if (!(await validProtocolPath(target))) {
    throw new Error("impact file must be below .gleam_mutants");
  }
  const temporary = `${target}.tmp-${Date.now()}-${Math.random().toString(16).slice(2)}`;
  const path = await import("node:path");
  try {
    if (globalThis.Deno) {
      await Deno.mkdir(path.dirname(target), { recursive: true });
      await Deno.writeTextFile(temporary, source);
      await Deno.rename(temporary, target);
      return;
    }
    const fs = await import("node:fs/promises");
    await fs.mkdir(path.dirname(target), { recursive: true });
    await fs.writeFile(temporary, source, "utf8");
    await fs.rename(temporary, target);
  } catch (error) {
    try {
      if (globalThis.Deno) await Deno.remove(temporary);
      else await (await import("node:fs/promises")).unlink(temporary);
    } catch (_) {
      // Preserve the atomic-write failure; cleanup is best effort.
    }
    throw error;
  }
}

async function writeImpact(descriptors, complete) {
  const target = environment("GLEAM_MUTANTS_TEST_IMPACT_FILE");
  if (!target) return;
  const impacts = globalThis[impactSymbol] instanceof Map
    ? globalThis[impactSymbol]
    : new Map();
  const reaches = descriptors.map(descriptor => ({
    test_id: descriptor.test_id,
    mutant_ids: [...(impacts.get(descriptor.test_id) ?? new Set())].sort(),
  }));
  await atomicWrite(target, JSON.stringify({
    schema_version: 1,
    runner: "smartest",
    runtime: runtimeName(),
    complete,
    tests: descriptors,
    reaches,
  }));
}

export async function main() {
  const protocolEnabled = Boolean(
    environment("GLEAM_MUTANTS_TEST_IMPACT_FILE")
      || environment("GLEAM_MUTANTS_TEST_SELECTION_FILE"),
  );
  if (protocolEnabled) delete globalThis[impactSymbol];
  let selectors;
  try {
    selectors = await loadSelection();
  } catch (error) {
    print(`FAIL test selection protocol\n${String(error?.message ?? error)}\n`);
    finish(2);
    return;
  }
  const package_ = await packageName();
  const filter = environment("SMARTEST_FILTER");
  let passed = 0;
  let failed = 0;
  let bindings = [];
  let manifestComplete = true;
  const results = [];
  const descriptors = [];
  const matchedSelectors = new Set();

  for await (const path of gleamFiles("test")) {
    const moduleName = path.slice("test/".length, -".gleam".length);
    const moduleUrl = new URL(`../../${moduleName}.mjs`, import.meta.url);
    const module = await import(moduleUrl.href);
    for (const functionName of Object.keys(module).sort()) {
      if (!functionName.endsWith("_test")) continue;
      if (!selected(moduleName, functionName, filter)) continue;
      const root = rootTestId(package_, moduleName, functionName);
      if (!selectedExport(root, selectors)) continue;
      try {
        const value = await with_test_context(root, () => module[functionName]());
        if (value !== undefined && value !== null && metadata(value) instanceof Metadata) {
          const entry = runner.entry(package_, moduleName, functionName, value);
          const ids = protocolEnabled
            ? Array.from(runner.test_ids(entry), evidence.test_id_to_string)
            : [];
          if (protocolEnabled) {
            descriptors.push(...ids.map(id => ({
              selector: id, test_id: id, kind: "smartest-leaf",
            })));
          }
          bindings.push(...runner.generator_bindings(entry));
          const replayId = environment("SMARTEST_REPLAY_ID");
          const options = replayId
            ? runner.replay_options(".", replayId, Date.now())
            : runner.workspace_options(".", Date.now());
          const matching = selectors === null
            ? ids
            : ids.filter(id => selectors.includes(id));
          for (const id of matching) matchedSelectors.add(id);
          const runResults = selectors === null
            ? runner.run_entry(entry, options)
            : runner.run_entry_selected(entry, options, toList(matching));
          for (const result of runResults) {
            results.push(result);
            if (runner.succeeded(result)) {
              print(".");
              passed += 1;
            } else {
              print(`\n${runner.render_result(result)}`);
              failed += 1;
            }
          }
        } else {
          if (protocolEnabled) {
            descriptors.push({
              selector: root, test_id: root, kind: "legacy-export",
            });
            matchedSelectors.add(root);
          }
          print(".");
          passed += 1;
          results.push(runner.legacy_result(
            package_, moduleName, functionName, true, "opaque legacy test",
          ));
        }
      } catch (error) {
        if (protocolEnabled) {
          descriptors.push({
            selector: root, test_id: root, kind: "legacy-export",
          });
          matchedSelectors.add(root);
        }
        const message = String(error?.stack ?? error);
        print(`\nFAIL ${moduleName}/${functionName}\n${message}\n`);
        failed += 1;
        manifestComplete = false;
        results.push(runner.legacy_result(
          package_, moduleName, functionName, false, message,
        ));
      }
    }
  }
  if (selectors !== null) {
    const unknown = selectors.filter(selector => !matchedSelectors.has(selector));
    if (unknown.length > 0) {
      print(`\nFAIL unknown test selector\n${JSON.stringify(unknown)}\n`);
      failed += 1;
    }
  }
  const reportResult = evidenceReport.write(".", new runner.Report(toList(results)));
  if (!reportResult.isOk()) {
    print(`\nFAIL evidence report\n${reportResult[0]}\n`);
    failed += 1;
  }
  if (selectors === null && !filter && manifestComplete) {
    const stored = storage.write_generator_manifest(".", toList(bindings));
    if (!stored.isOk()) {
      print(`\nFAIL generator manifest\n${stored[0]}\n`);
      failed += 1;
    }
  } else if (selectors === null && !filter && !manifestComplete) {
    print("\nGenerator manifest was not replaced because discovery was incomplete.\n");
  }
  try {
    await writeImpact(
      descriptors,
      // A filter belongs to this invocation's configured suite. Full-suite
      // confirmations run with the same environment and therefore the same
      // filter.
      manifestComplete && selectors === null,
    );
  } catch (error) {
    print(`\nFAIL test impact manifest\n${String(error?.message ?? error)}\n`);
    failed += 1;
  }
  print(`\n${passed} passed, ${failed} failed\n`);
  finish(failed === 0 ? 0 : 1);
}
